src/Components/InfiniteScroll.tsx

```diff
@@ -103,232 +99,290 @@ const DefaultContentPlaceHolder = ({ height }: IContentPlaceHolderProps) => {
-        // Initialize page height map and estimated page height
-        this.props.updatePageHeightMap(this.props.query);
+      if (this.props.query.response) {
+        this.props.updateStateFromQuery(this.props.query);
       }
     },
-    componentWillUnmount() {
-      this.props.handleScrollDebounced.cancel();
-      this.props.handleScrollThrottled.cancel();
+    componentDidUpdate(prevProps) {
+      if (prevProps.query !== this.props.query && this.props.query.response) {
+        this.props.updateStateFromQuery(this.props.query);
+      } else if (this.props.currentItems && this.props.pageHeightMap === prevProps.pageHeightMap) {
+        this.props.updatePageHeightMap();
+      }
     },
   }),
+  withLoadingIndicator({
+    isLoading: ({ currentItems }) => currentItems == null,
+    loadingComponent: LoadingComponent
+  }),
   omitProps([
-    "debounceTime",
-    "handleScroll",
-    "handleScrollDebounced",
-    "handleScrollThrottled",
+    "currentPage",
+    "items",
     "lastScrollTop",
-    "recordPageHeight",
-    "updateQuery",
+    "loadingComponent",
+    "loadingComponentProps",
+    "totalPages",
     "updatePageHeightMap",
+    "updateState",
+    "updateStateFromQuery",
   ]),
-  pure,
 )(({
   containerComponent: ContainerComponent,
   contentPlaceHolderComponent: ContentPlaceHolder,
+  contentPlaceHolderComponentProps,
+  currentItems,
+  currentPageEndMarkerRef,
+  currentPageStartMarkerRef,
   disableVirtualScrolling,
   estimatedPageHeight,
-  handleScrollPersistingEvent,
+  handleScroll,
+  hasPreviousPage,
+  hasNextPage,
   modelComponent: ModelComponent,
   modelComponentProps,
   modelName,
-  nextPageHeightRef,
+  nextPage,
+  nextPageQuery,
   nextPlaceHolderHeight,
   pageHeightMap,
-  previousPlaceHolderHeight,
-  previousPageStartMarkerRef,
+  previousPage,
+  previousPageQuery,
   previousPageEndMarkerRef,
-  currentPageStartMarkerRef,
-  currentPageEndMarkerRef,
+  previousPageStartMarkerRef,
+  previousPlaceHolderHeight,
   nextPageStartMarkerRef,
   nextPageEndMarkerRef,
-  query: queryManager,
   ...containerProps
 }) => (
-  <ContainerComponent {...containerProps} onScroll={handleScrollPersistingEvent}>
-    {disableVirtualScrolling && (
-      <DisplayPreviousPage queryManager={queryManager} modelComponent={ModelComponent} modelComponentProps={modelComponentProps} modelName={modelName} />
+  <ContainerComponent {...containerProps} onScroll={handleScroll}>
+    {hasPreviousPage && (
+      disableVirtualScrolling
+        ? (
+          <InfiniteScrollPreviousPage
+            modelComponent={ModelComponent}
+            modelComponentProps={modelComponentProps}
+            query={previousPageQuery}
+          />
+        ) : (
+          <>
+            <ContentPlaceHolder
+              height={previousPlaceHolderHeight}
+              {...contentPlaceHolderComponentProps}
+            />
+
+            <script ref={previousPageStartMarkerRef} />
```

Comment:
This is new to me.  I will have to take a closer look at the docs.

Reply:
The use of script tags was approved previously (already sitting in master). It's a trick I picked up after reviewing multiple infinite scroll / virtual scroll solutions a few months ago before we started the Infinite Scroll epic.

Essentially, we need to know the height of a page, but we cannot add extra html elements to the DOM as it causes issues. For example, these script tags were introduced because we previously had `<span>` tags wrapping each page, and that was causing tables to break, because it was putting `<span>` tags directly inside the `<tbody>` around the `<tr>` tags, which is not html valid and would cause React to throw an error.

Reply:
Oh, by "this," I just meant the ref stuff.

Reply:
Ahh. Yeah it's a pretty cool feature of React. Basically it hands you a pointer to the actual element in the DOM. https://reactjs.org/docs/refs-and-the-dom.html